docs: Idempotent runs V0 documentation (#819)
kurtosis-tech/kurtosis: docs/docs/explanations/starlark-idempotent-runs.md

---

> title: Starlark idempotent runs

Reviewer: ```suggestion
title: Idempotent runs
```

> sidebar_label: Starlark idempotent runs

Reviewer: ```suggestion
sidebar_label: Idempotent runs
```

---

Background
----------

> Did you notice those lines in the `kurtosis run` command output?

Reviewer: ```suggestion
When running the `kurtosis run` command, you may notice the following message get printed:
```

```console
SKIPPED - This instruction has already been run in this enclave
```

> To reduce execution time, Kurtosis tries to optimize each run of a Starlark package based on what has already been run 

Reviewer: ```suggestion
The reason this happens is because Kurtosis will optimize each run of a Starlark package based on what has already been run in a given enclave, thus reducing execution time and resources.
```

> in the enclave. 

Reviewer: ```suggestion
```

> For example if you try to run twice the exact same package, Kurtosis will skip all the instructions for the second run

Reviewer: ```suggestion
This means when you try to run the exact same package twice in a row, Kurtosis will skip all the instructions for the second run because they were already executed in the first run.
```

> because they were already executed in the first run. 

Reviewer: ```suggestion
```

:::info

> This feature is still experimental and can be deactivated by adding `--experimental NO_INSTRUCTIONS_CACHING` parameter 

Reviewer: ```suggestion
This feature is still experimental and can be deactivated by adding `--experimental NO_INSTRUCTIONS_CACHING` parameter to the `kurtosis run` command.
```

> to the `kurtosis run` command.

Reviewer: ```suggestion
```

:::

How it works
------------

#### Definitions

> We call the _enclave plan_ the sequence of Starlark instructions that were executed inside this enclave.

Reviewer: ```suggestion
The __enclave plan__ is defined as the sequence of Starlark instructions that were previously executed inside a given enclave. Meanwhile, the __submitted plan__ is defined as the set of instructions generated by interpreting the package before it gets executed.

Within a plan, whether it be an _enclave_ or _submitted_ plan, Kurtosis considers a given instruction `instruction_2` to be dependent on another instruction `instruction_1` if and only if `instruction_2` is sequenced to take place _after_ `instruction_1` in the plan's instruction sequence.
```

> When a new package is run, we call the _submitted plan_ the set of instructions generated by interpreting the package but not yet executed.

Reviewer: ```suggestion
```

> Inside a plan (enclave plan or submitted plan), it is said that an instruction (i.e. `instruction_1`) _depends_ on 

Reviewer: ```suggestion
```

> another instruction (`instruction_2`) iff `instruction_2` is located _before_ `instruction_1` in the plan's instructions

Reviewer: ```suggestion
```

> sequence.

Reviewer: ```suggestion
```

> Two instructions are said to be _equivalent_ iff their Starlark representations are strictly equal.

Reviewer: ```suggestion
Two instructions are said to be _equivalent_ if and only if their Starlark representations are strictly equal.
```

#### Criteria to skip an instruction

> The way Kurtosis optimize the submitted plan before executing it is by following this heuristic:

Reviewer: ```suggestion
Kurtosis uses the following heuristic to determine which parts of a plan to skip in order to optimize runs:
```

> 1. any instruction from the _submitted plan_ that is executed inside the enclave is appended to the _enclave plan_

Reviewer: ```suggestion
1. Any instruction from the _submitted plan_ that is executed inside the enclave is appended to the _enclave plan_
```

> 1. for any given instruction from the _submitted plan_, it is effectively skipped (i.e. not executed) if:

Reviewer: ```suggestion
1. For any given instruction from the _submitted plan_, it is effectively skipped (i.e. not executed) if:
```

>    1. it exists one instruction in the *enclave plan* equivalent to this instruction (called the _matched instruction_ below)

Reviewer: ```suggestion
   1. There exists one instruction in the *enclave plan* that is equivalent to this instruction in the _enclave plan_(called the _matched instruction_ below)
```

>    1. and ONE of the following assertions is true:

Reviewer: ```suggestion
   1. And if at least ONE of the following assertions is true:
```

>       1. there is no direct dependent instruction for this instruction in the _submitted plan_. Put it simply, it is the last instruction of the _submitted plan_

Reviewer: ```suggestion
      1. There is no direct dependent instruction for the given instruction in the _submitted plan_. Put it simply, it is the last instruction of the _submitted plan_.
```

>       1. there is no direct dependent instruction for the _matched instruction_ in the _enclave plan_. Put it simply, the _matched instruction_ is the last instruction of the _enclave plan_

Reviewer: ```suggestion
      1. There is no direct dependent instruction for the _matched instruction_ in the _enclave plan_. Put it simply, the _matched instruction_ is the last instruction of the _enclave plan_.
```

>       1. the sequence of direct dependent instructions for the _matched instruction_ in the _enclave plan_ is a _prefix_ of the sequence of direct dependent instructions of the _instruction_ from the _submitted plan_

Reviewer: ```suggestion
      1. The sequence of direct dependent instructions for the _matched instruction_ in the _enclave plan_ is a _prefix_ of the sequence of direct dependent instructions of the _instruction_ from the _submitted plan_.
```
